server/src/main/java/org/elasticsearch/common/logging/NodeAndClusterIdConverter.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * Licensed to Elasticsearch under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.elasticsearch.common.logging;
+
+import org.apache.logging.log4j.core.LogEvent;
+import org.apache.logging.log4j.core.config.plugins.Plugin;
+import org.apache.logging.log4j.core.pattern.ConverterKeys;
+import org.apache.logging.log4j.core.pattern.LogEventPatternConverter;
+import org.apache.logging.log4j.core.pattern.PatternConverter;
+
+import java.util.Locale;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Pattern converter to format the node_and_cluster_id variable into a json fields node.id and cluster.uuid
+ * Keeping those two fields together assures that the will be atomically set and become visible in logs at the same time
+ */
+@Plugin(category = PatternConverter.CATEGORY, name = "NodeAndClusterIdConverter")
+@ConverterKeys({"node_and_cluster_id"})
+public final class NodeAndClusterIdConverter extends LogEventPatternConverter {
+    private static final AtomicReference<String> nodeAndClusterIdsReference = new AtomicReference<>();
```

Comment:
The plural in the name `nodeAndClusterIdsReference` implies that there are multiple node and cluster ids when there is only one. Also I think the suffix `reference` does not add meaning? So maybe just `nodeAndClusterId`? Also, do we want to declare this as plain final field instead of a class constant given that `NodeAndClusterIdConverter` is effectively a singleton (as per Log4j's contract)?

Reply:
agree the name change here would be beneficial - done

re - plain final field. Relying on the  fac that Log4j only instantiates this once makes it easier to pass the value from `NodeAndClusterIdStateListener.java` to this field with the use of static setOnce method.  
If this was a regular final field the cluster state listener would have to access that only instance used by log4j - therefore this class would have to implement a `newInstance` in a singleton style. This is how it was done previously and I think it was redundant.